pyhanko/sign/signers/csc_signer.py

```diff
@@ -0,0 +1,155 @@
+        """
+        self.use_pades = use_pades
+        self.timestamper = timestamps.HTTPTimeStamper(tsa_url)
+        certs_encoded = csc_credentials_info_response['cert/certificates']
+        certs = [x509.Certificate.load(b64decode(cert)) for cert in certs_encoded]
+        cert_registry = SimpleCertificateStore()
+        cert_registry.register_multiple(set(certs))
+        super().__init__(
+            signing_cert=certs[0],
+            cert_registry=cert_registry,
+            signature_mechanism=None,
+            signature_value=signature_value,
+        )
+
+    def prep_document(self, pdf: bytes, config: TPrepDocumentConfig):
+        """
+        This method is pdf specific.
+        It returns a pdf document that is ready to have a CMS attached.
+        Also it computes the pdf specific hash of the prepared document.
+        """
+        pdf_signer = signers.PdfSigner(
+            signers.PdfSignatureMetadata(
+                field_name='SigNew',
+                embed_validation_info=config.get("embed_validation_info", False),
+                use_pades_lta=self.use_pades,
+                subfilter=config.get("subfilter", SigSeedSubFilter.ADOBE_PKCS7_DETACHED),
+                validation_context=config.get("validation_context", as_any(None)),
+                md_algorithm=config.get('md_algorithm', 'sha256'),
+            ),
+            signer=self,
+            timestamper=self.timestamper,
+        )
+        input_buf = BytesIO(pdf)
+        pdf_out = IncrementalPdfFileWriter(input_buf)
+        prep_digest, tbs_document, output = pdf_signer.digest_doc_for_signing(pdf_out)
+
+        return {
+            "digest": {
+                "data_digest": prep_digest.document_digest,
+                "digest_algorithm": prep_digest.md_algorithm,
+            },
+            "pdf": {
+                "prep_digest": prep_digest,
+                "post_sign_instructions": tbs_document.post_sign_instructions,
+                "output": output,
+            },
+        }
+
+    def generate_signed_attrs(self, digest: TDigest):
+        """
+        Get payload to be signed. This method is not merged with "prep_document" to:
+        - allow hash signing without having the document;
+        - be pdf agnostic.
+        """
+        return self.signed_attrs(
+            digest["data_digest"], digest['digest_algorithm'], use_pades=self.use_pades
+        )
+
+    def generate_cms(self, digest: TDigest, sig_value: bytes = None):
+        """
+        Generate cms from signed payload. The sig_value should be the PKCS1
+        signature returned by CSC signatures/signHash endpoint. (The CSC spec does
+        not state that signatures are PKCS1 but the example response suggests that it is.)
+        """
+        if sig_value:
+            self._signature_value = sig_value
+        if not any(self._signature_value):  # all bytes are zeroes
+            raise Exception("Fake signature value is used.")
```

Comment:
It would seem that this sentinel value should go into the base class `ExternalSigner`. And it seems the `dry_run` flag serves a similar function.

Maybe all this erroring should be done in the `sign_raw` function. Like this:
```
class ExternalSigner(Signer):                                         
    def sign_raw(self, data: bytes, digest_algorithm: str, dry_run=False) \
            -> bytes:
        """
        Return a fixed signature value.
        """
        if dry_run == False and not all(self._signature_value):
            raise SigningError("Fake signature value is used when signing for real.")
        return self._signature_value
```

Reply:
This does not address the initial problem, but maybe is good enough?

Reply:
Yup, I'm OK with this way of solving it.